### PR TITLE
add success message box with commit hash

### LINKED ISSUE #3560: Show commit SHA at successfull commit
Is it possible to show the commit SHA in the message that the commit is done?

A link would be even better but that probably needs some customizing depending what is in the backend (GitHub, GitLab, Bitbucket etc...)

### PR DESCRIPTION
**Pull request by bigld**
fixes #3560 

The commit hash is shown in a message box. Link to the commit is prepared but not yet implemented
image

### CODE REVIEW

**Comment by larshp** [comment]
@sbcgua any comments for the html/css?

File: src/ui/zcl_abapgit_gui_page.clas.abap [hunk 1]
```diff
@@ -19,10 +19,12 @@ CLASS zcl_abapgit_gui_page DEFINITION PUBLIC ABSTRACT
         page_menu  TYPE REF TO zcl_abapgit_html_toolbar,
       END OF  ty_control .
 
-    DATA ms_control TYPE ty_control .
+    CLASS-DATA gs_success_log_entry TYPE zif_abapgit_log=>ty_log_out.
```
Comment (sbcgua) [inline src/ui/zcl_abapgit_gui_page.clas.abap:22]: Mmmm, this supposes subclassing of gui_page. And currently some refactoring are done to remove the subclassing and make gui_page final. Also this is a global variable in fact which is not good. Maybe make sense to do via deferred_parts ... or maybe leave the feature without html green banner - see the main review comment

File: src/ui/zcl_abapgit_gui_page_commit.clas.abap [hunk 2]
```diff
@@ -479,7 +483,14 @@ CLASS ZCL_ABAPGIT_GUI_PAGE_COMMIT IMPLEMENTATION.
...
 
-        MESSAGE 'Commit was successful' TYPE 'S' ##NO_TEXT.
+        lv_repo_log = mo_repo->get_log( ).
+        lt_log_messages = lv_repo_log->get_messages( ).
+        READ TABLE lt_log_messages WITH KEY obj_type = 'SHA1' ASSIGNING <ls_message>.
+        IF sy-subrc = 0.
+          gs_success_log_entry = <ls_message>.
+        ELSE.
+          MESSAGE 'Commit was successful' TYPE 'S' ##NO_TEXT.
+        ENDIF.
```
Comment (sbcgua) [inline src/ui/zcl_abapgit_gui_page_commit.clas.abap:493]: Should not it be one unified approach to show messages ? Html only or message only ?

File: src/zcl_abapgit_repo.clas.abap [hunk 3]
```diff
@@ -500,6 +505,16 @@ CLASS ZCL_ABAPGIT_REPO IMPLEMENTATION.
   ENDMETHOD.
 
 
+  METHOD get_or_create_log.
```
Comment (sbcgua) [inline src/zcl_abapgit_repo.clas.abap:508]: Can this be merged into get_log ? 

P.S. @larshp could you pls have a look on this too. I didn't see into repo code for quite a while . Also I rememeber there was an intention to remove log from repo at all.

**Comment by bigld** [comment]
The basic idea is to have a link to jump to jump to commit/branch/... of the online repo provider, so html would be cool. 
And maybe there will be further stuff like create PR, URL to clipboard,.... 

Additionally I like the idea of actions for (error) messages and get rid of SAP GUI MESSAGE...

It would also make sense to have a global/general message container to hold several info/success/warning messages - some kind of current log function. 

But start small...

**Comment by sbcgua** [comment]
> The basic idea is to have a link to jump to jump to commit/branch/

Why not adding a link to the last commit at the repo header level ? The message will disappear ... 

image

btw the "disappearing" is a weak point of html messages - e.g. with errors I can do some actions at the page and it does not disappear - this is very unintuitive and confusing (although the tech itself is very cool)

**Comment by bigld** [comment]
OK - I will remove HTML message and show the hash in SAP message. and also adjust the repo log handling a bit

**Comment by bigld** [comment]
ZCL_ABAPGIT_LOG now use multiton for repo logs.
success message now shows commit hash as standard message
image

**Comment by bigld** [comment]
image

File: src/ui/zabapgit_css_common.w3mi.data.css [hunk 4]
```diff
@@ -836,6 +836,12 @@ div.message-panel:hover .message-panel-commands {
   display: block
 }
 
+/* Success message Panel */
+div.success-panel {
+  border-color: hsl(148, 87%, 33%);
+  background-color: hsla(148, 67%, 92%, 1);
+}
```
Comment (sbcgua) [inline src/ui/zabapgit_css_common.w3mi.data.css:843]: Is this change still relevant ?

File: src/ui/zcl_abapgit_gui_chunk_lib.clas.abap [hunk 5]
```diff
@@ -116,6 +116,18 @@ CLASS zcl_abapgit_gui_chunk_lib DEFINITION
...
 
   PROTECTED SECTION.
+    CLASS-METHODS render_repo_top_commit_hash
+      IMPORTING
+        iv_html        TYPE REF TO zcl_abapgit_html
+        iv_repo_online TYPE REF TO zcl_abapgit_repo_online
+        iv_repo_url    TYPE zif_abapgit_persistence=>ty_repo-url
+      RAISING
+        zcx_abapgit_exception.
+    CLASS-METHODS get_commit_display_url
```
Comment (sbcgua) [inline src/ui/zcl_abapgit_gui_chunk_lib.clas.abap:126]: Suggest to move it to zcl_abapgit_utils or some other utils. The chunk lib is about rendering, and this method is a "backend" calculation.

File: src/ui/zcl_abapgit_gui_page.clas.abap [hunk 6]
```diff
@@ -17,12 +17,12 @@ CLASS zcl_abapgit_gui_page DEFINITION PUBLIC ABSTRACT
       BEGIN OF ty_control,
         page_title TYPE string,
         page_menu  TYPE REF TO zcl_abapgit_html_toolbar,
-      END OF  ty_control .
+      END OF  ty_control.
```
Comment (sbcgua) [inline src/ui/zcl_abapgit_gui_page.clas.abap:20]: Are the changes in gui_page still relevant ?

File: src/utils/zcl_abapgit_log.clas.abap [hunk 7]
```diff
@@ -37,6 +47,36 @@ ENDCLASS.
...
+    DATA: ls_repo_log TYPE ty_repo_log_instance.
+    FIELD-SYMBOLS: <ls_repo_log> TYPE ty_repo_log_instance.
+
+    READ TABLE gt_repo_logs ASSIGNING <ls_repo_log> WITH KEY repo_key = iv_repo_key.
+    IF sy-subrc <> 0.
+      ls_repo_log-repo_key = iv_repo_key.
+      CREATE OBJECT ls_repo_log-instance TYPE zcl_abapgit_log EXPORTING iv_title = iv_log_title.
+      INSERT ls_repo_log INTO TABLE gt_repo_logs ASSIGNING <ls_repo_log>.
+    ELSEIF iv_log_title IS SUPPLIED.
+      <ls_repo_log>-instance->set_title( iv_log_title ).
```
Comment (sbcgua) [inline src/utils/zcl_abapgit_log.clas.abap:73]: Mmmm ... singleton instance ? Could it already have some content ? and the the title redefined ? and what happens to the content ? Not really clear.

File: src/zif_abapgit_definitions.intf.abap [hunk 8]
```diff
@@ -518,6 +518,7 @@ INTERFACE zif_abapgit_definitions
       changed_by                    TYPE string VALUE 'changed_by',
       documentation                 TYPE string VALUE 'documentation',
       changelog                     TYPE string VALUE 'changelog',
+      goto_repo                     TYPE string VALUE 'goto_repo',
```
Comment (sbcgua) [inline src/zif_abapgit_definitions.intf.abap:521]: Is this relevant ? Seem to not used in the rest of the code

**Comment by bigld** [comment]
I'll close this PR and split it into 3 smaller ones

**Pull request closed without merging**